Clean/Utils/Bitwise.lean

```diff
@@ -0,0 +1,102 @@
+import Mathlib.Tactic
+import Mathlib.Algebra.Field.ZMod
+import Clean.Utils.Field
+
+namespace Bytes
+def split_two_bytes (i : Fin (256 * 256)) : Fin 256 × Fin 256 :=
+  let x := i.val / 256
+  let y := i.val % 256
+  have x_lt : x < 256 := by simp [x, Nat.div_lt_iff_lt_mul]
+  have y_lt : y < 256 := Nat.mod_lt i.val (by norm_num)
+  (⟨ x, x_lt ⟩, ⟨ y, y_lt ⟩)
+
+def concat_two_bytes (x y : Fin 256) : Fin (256 * 256) :=
+  let i := x.val * 256 + y.val
+  have i_lt : i < 256 * 256 := by
+    unfold i
+    linarith [x.is_lt, y.is_lt]
+  ⟨ i, i_lt ⟩
+
+theorem concat_split (x y : Fin 256) : split_two_bytes (concat_two_bytes x y) = (x, y) := by
+  dsimp [split_two_bytes, concat_two_bytes]
+  ext
+  simp only
+  rw [mul_comm]
+  have h := Nat.mul_add_div (by norm_num : 256 > 0) x.val y.val
+  rw [h]
+  simp
+  simp only [Nat.mul_add_mod_of_lt y.is_lt]
+
+section
+variable {p : ℕ}  [Fact p.Prime] [p_large_enough: Fact (p > 512)]
+
+def from_byte (x: Fin 256) : F p :=
+  FieldUtils.nat_to_field x.val (by linarith [x.is_lt, p_large_enough.elim])
+
+lemma from_byte_lt (x: Fin 256) : (from_byte (p:=p) x).val < 256 := by
+  dsimp [from_byte]
+  rw [FieldUtils.val_of_nat_to_field_eq]
+  exact x.is_lt
+
+lemma from_byte_eq (x : F p) (x_lt : x.val < 256) : from_byte ⟨ x.val, x_lt ⟩ = x := by
+  dsimp [from_byte]
+  apply FieldUtils.nat_to_field_of_val_eq_iff
+
+lemma from_byte_cast_eq {z: F p} (z_lt : z.val < 256) : from_byte z.cast = z := by
+  simp only [from_byte]
+  have : (z.cast : Fin 256).val = z.val := ZMod.val_cast_eq_val_of_lt z_lt
+  simp only [this]
+  apply FieldUtils.nat_to_field_of_val_eq_iff
+end
+end Bytes
+
+namespace Bitwise
```

Comment:
these are a couple new lemmas about bitwise operations needed for the new soundness proofs